### PR TITLE
fix(block-device): allow client to label and annotate bd

### PR DESCRIPTION
**Pull request by shovanmaity**
This PR fix metadata part of a block device. Now users can label and annotate block devices. What it does it actually takes metadata of older block device and patch that with system generated information.

Signed-off-by: Shovan Maity <shovan.maity@mayadata.io>

### CODE REVIEW

File: cmd/ndm_daemonset/controller/blockdevicestore.go [hunk 1]
```diff
@@ -93,7 +93,42 @@ func (c *Controller) UpdateBlockDevice(blockDevice apis.BlockDevice, oldBlockDev
...
+	// - selfLink - populated by the system we should use old object.
+	// - uid - populated by the system we should use old object.
+	// - resourceVersion - populated by the system we should use old object.
+	// - generation - populated by the system we should use old object.
+	// - creationTimestamp - populated by the system we should use old object.
+	// - deletionTimestamp - populated by the system we should use old object.
+	// - deletionGracePeriodSeconds - populated by the system we should use old object.
+	// - labels - we will patch older labels with new labels.
+	// - annotations - we will patch older annotations with new annotations.
+	// - ownerReferences TODO
```
Comment (shovanmaity) [inline cmd/ndm_daemonset/controller/blockdevicestore.go]: for TODO items we need to decide should we allow the client to add them. If not then we can assign them with system generated value

File: cmd/ndm_daemonset/controller/blockdevicestore.go [hunk 2]
```diff
@@ -93,7 +93,42 @@ func (c *Controller) UpdateBlockDevice(blockDevice apis.BlockDevice, oldBlockDev
...
+	// - initializers TODO
+	// - finalizers TODO
+	// - clusterName - no patch required we can use old object.
+	metadata := oldBlockDevice.ObjectMeta
+
+	// Patch older label with new label. If there is a new key then it will be added
+	// if it is an existing key then value will be overwritten with value from new label
+	if blockDeviceCopy.ObjectMeta.Labels != nil {
```
Comment (akhilerm) [inline cmd/ndm_daemonset/controller/blockdevicestore.go]: can this be made a separate private `func()`. I think the `Update` have now become too big to understand.

File: cmd/ndm_daemonset/controller/blockdevicestore_test.go [hunk 3]
```diff
@@ -168,9 +168,11 @@ func TestUpdateDevice(t *testing.T) {
...
 	// Retrieve disk resource
 	cdevR, err := fakeController.GetBlockDevice(fakeDeviceUID)
-
-	devR.ObjectMeta.Name = "disk-updated-fake-uuid"
-	err = fakeController.UpdateBlockDevice(devR, cdevR)
+	if err != nil {
+		t.Fatal(err)
+	}
+	cdevR.ObjectMeta.Name = "disk-updated-fake-uuid"
+	err = fakeController.UpdateBlockDevice(*cdevR, nil)
```
Comment (akhilerm) [inline cmd/ndm_daemonset/controller/blockdevicestore_test.go]: Why is this change required?

Comment (shovanmaity) [reply cmd/ndm_daemonset/controller/blockdevicestore_test.go]: From the error log -
`If the resource is not present then it should return error.`
My understanding - there maybe 2 cases -
1. If the resource is not present we can get error while fetching it.
2. If the resource is not present we can get if we try to update it.

Earlier it was returning error for case 2 as was getting name and namespace from new object. Now we change that we are taking name and namespace from the older object so it will not fail.
Now it can fail for case 1 only and for that, we need to pass the old object as nil.

File: cmd/ndm_daemonset/controller/blockdevicestore.go (hunk 2 above)
Comment (shovanmaity) [reply cmd/ndm_daemonset/controller/blockdevicestore.go]: I am ok with it.
Can you suggest what should that function do only merging metadata or full bd object

Comment (akhilerm) [reply cmd/ndm_daemonset/controller/blockdevicestore.go]: a function to patch objectMeta will be enough.
```
func mergeMetaData(oldBlockDevice.ObjectMeta, blockDeviceCopy.ObjectMeta) ObjectMeta {
   // All the todo comments
   // merging labels
   // merging annotations
   return metadata
}
```

File: cmd/ndm_daemonset/controller/blockdevicestore.go [hunk 4]
```diff
@@ -253,3 +253,40 @@ func (c *Controller) MarkBlockDeviceStatusToUnknown() {
 		}
 	}
 }
+
```
Comment (akhilerm) [inline cmd/ndm_daemonset/controller/blockdevicestore.go]: can you provide the function comments.

File: cmd/ndm_daemonset/controller/blockdevicestore.go [hunk 5]
```diff
@@ -253,3 +253,42 @@ func (c *Controller) MarkBlockDeviceStatusToUnknown() {
...
+	// - labels - we will patch older labels with new labels.
+	// - annotations - we will patch older annotations with new annotations.
+	// - ownerReferences as ndm-ds is not adding ownerReferences we can go with old object.
+	// - initializers ^^^
+	// - finalizers ^^^
+	// - clusterName - no patch required we can use old object.
+	metadata := oldMetadata
+	// Patch older label with new label. If there is a new key then it will be added
+	// if it is an existing key then value will be overwritten with value from new label
+	if newMetadata.Labels != nil {
```
Comment (sonasingh46) [inline cmd/ndm_daemonset/controller/blockdevicestore.go]: Do we need this nil check?

File: cmd/ndm_daemonset/controller/blockdevicestore.go [hunk 6]
```diff
@@ -253,3 +253,42 @@ func (c *Controller) MarkBlockDeviceStatusToUnknown() {
...
+		for key, value := range newMetadata.Labels {
+			metadata.Labels[key] = value
+		}
+	}
+
+	// Patch older annotations with new annotations. If there is a new key then it will be added
+	// if it is an existing key then value will be overwritten with value from new annotations
+	if newMetadata.Annotations != nil {
```
Comment (sonasingh46) [inline cmd/ndm_daemonset/controller/blockdevicestore.go]: Similarly, do we need this nil check?

File: cmd/ndm_daemonset/controller/blockdevicestore.go [hunk 7]
```diff
@@ -253,3 +253,42 @@ func (c *Controller) MarkBlockDeviceStatusToUnknown() {
...
+	// - creationTimestamp - populated by the system we should use old object.
+	// - deletionTimestamp - populated by the system we should use old object.
+	// - deletionGracePeriodSeconds - populated by the system we should use old object.
+	// - labels - we will patch older labels with new labels.
+	// - annotations - we will patch older annotations with new annotations.
+	// - ownerReferences as ndm-ds is not adding ownerReferences we can go with old object.
+	// - initializers ^^^
+	// - finalizers ^^^
+	// - clusterName - no patch required we can use old object.
+	metadata := oldMetadata
```
Comment (sonasingh46) [inline cmd/ndm_daemonset/controller/blockdevicestore.go]: Why we need a new `metadata` variable ? Can't we just use `oldMetadata`
